query/src/main/java/org/infinispan/search/mapper/work/impl/SearchIndexerImpl.java

```diff
@@ -1,39 +1,112 @@
 
 /**
  * @author Fabio Massimo Ercoli
  */
 public class SearchIndexerImpl implements SearchIndexer {
+   private static final Log log = LogFactory.getLog(QueryInterceptor.class, Log.class);
 
    private final PojoIndexer delegate;
    private final EntityConverter entityConverter;
    private final InfinispanTypeContextProvider typeContextProvider;
 
-   private final SimpleScheduledExecutor offloadingExecutor;
+   private final int maxConcurrency;
+   private final ConcurrentMap<Supplier<Flowable<?>>, CompletableFuture<?>> submittedTasks;
+
+   private Disposable processorDisposer;
+
+   private FlowableProcessor<Supplier<Flowable<?>>> requestProcessor;
 
    public SearchIndexerImpl(PojoIndexer delegate, EntityConverter entityConverter,
-         InfinispanTypeContextProvider typeContextProvider, BlockingManager blockingManager) {
+                            InfinispanTypeContextProvider typeContextProvider, int maxConcurrency) {
       this.delegate = delegate;
       this.entityConverter = entityConverter;
       this.typeContextProvider = typeContextProvider;
-      this.offloadingExecutor = InfinispanIndexingExecutorProvider.writeExecutor(blockingManager);
+      this.maxConcurrency = maxConcurrency;
+
+      this.submittedTasks = new ConcurrentHashMap<>();
+   }
+
+   @Override
+   public void start() {
+      requestProcessor = UnicastProcessor.<Supplier<Flowable<?>>>create().toSerialized();
+      // onBackpressureLatest will drop any item that it can't immediately pass downstream
+      // Thus next downstream must batch them to not immediately drop
+      processorDisposer = requestProcessor.onBackpressureLatest(supplier -> {
+               CompletableFuture<?> completableFuture = submittedTasks.remove(supplier);
+               if (completableFuture == null) {
+                  throw new IllegalStateException("Dropped task " + supplier + " not found in submittedTask " + submittedTasks);
+               }
+               completableFuture.completeExceptionally(log.hibernateSearchBackpressure());
+            })
+            // We give HS some extra buffer room before we start throwing back pressure exceptions if
+            // the flatMap operations can't keep up
+            .rebatchRequests(10_000)
+            // This will only request up to maxConcurrency items at the same time
+            .flatMap(Supplier::get, maxConcurrency)
+            // Clear the submittedTasks on error/completion/cancel just in case
+            .doFinally(submittedTasks::clear)
+            .subscribe(Functions.emptyConsumer(),
+                  t -> log.fatal("Indexing encountered a non recoverable error", t));
+   }
+
+   private CompletableFuture<?> sendOperation(Supplier<CompletionStage<?>> actualSupplier) {
+      CompletableFuture<Object> future = new CompletableFuture<>();
+
+      Supplier<Flowable<?>> supplier = () -> {
+         try {
+            actualSupplier.get()
+                  .whenComplete((v, t) -> {
```

Comment:
This needs to be invoked in a non blocking thread! Otherwise we are leaking the blocking threads to ISPN application code since this supplier is invoked in a blocking thread.

Reply:
Resolved by adding a method to BlockingManager to get the NonBlockingExecutor for use here.